src/expire.c

```diff
@@ -385,7 +385,8 @@ void expireSlaveKeys(void) {
                     activeExpireCycleTryExpire(server.db+dbid,expire,start))
                 {
                     expired = 1;
-                    /* DELs aren't propagated, but modules may want their hooks. */
+                    /* Propagate the DEL (writable replicas do not propagate anything to other replicas,
+                     * but they might propagate to AOF) and trigger module hooks. */
                     postExecutionUnitOperations();
```

Comment:
@judeng can you please try to comment out this line, build, and see that indeed your test triggers that assertion?

Reply:
sure, I used 7.0.8 for this test, and always trigger a crash
```
jud@DESKTOP-4VAMVGD:~/judeng/redis$ tclsh tests/test_helper.tcl --single unit/expire
Cleanup: may take some time... OK
Starting test server at port 21079
[ready]: 12847
Testing unit/expire
[ready]: 12848
[ready]: 12846
[ready]: 12849
[ready]: 12850
[ready]: 12851
[ready]: 12852
[ready]: 12853
[ready]: 12854
[ready]: 12855
[ready]: 12857
[ready]: 12856
[ready]: 12859
[ready]: 12858
[ready]: 12860
[ready]: 12861
[ok]: EXPIRE - set timeouts multiple times (3 ms)
[ok]: EXPIRE - It should be still possible to read 'x' (1 ms)
[ok]: EXPIRE - After 2.1 seconds the key should no longer be here (2101 ms)
[ok]: EXPIRE - write on expire should work (2 ms)
[ok]: EXPIREAT - Check for EXPIRE alike behavior (1 ms)
[ok]: SETEX - Set + Expire combo operation. Check for TTL (0 ms)
[ok]: SETEX - Check value (1 ms)
[ok]: SETEX - Overwrite old key (0 ms)
[ok]: SETEX - Wait for the key to expire (1101 ms)
[ok]: SETEX - Wrong time parameter (0 ms)
[ok]: PERSIST can undo an EXPIRE (1 ms)
[ok]: PERSIST returns 0 against non existing or non volatile keys (1 ms)
[ok]: EXPIRE precision is now the millisecond (2003 ms)
[ok]: PSETEX can set sub-second expires (102 ms)
[ok]: PEXPIRE can set sub-second expires (104 ms)
[ok]: PEXPIREAT can set sub-second expires (204 ms)
[ok]: TTL returns time to live in seconds (1 ms)
[ok]: PTTL returns time to live in milliseconds (1 ms)
[ok]: TTL / PTTL / EXPIRETIME / PEXPIRETIME return -1 if key has no expire (2 ms)
[ok]: TTL / PTTL / EXPIRETIME / PEXPIRETIME return -2 if key does not exit (1 ms)
[ok]: EXPIRETIME returns absolute expiration time in seconds (0 ms)
[ok]: PEXPIRETIME returns absolute expiration time in milliseconds (0 ms)
[ok]: Redis should actively expire keys incrementally (606 ms)
[ok]: Redis should lazy expire keys (1005 ms)
[ok]: EXPIRE should not resurrect keys (issue #1026) (1002 ms)
[ok]: 5 keys in, 5 keys out (2 ms)
[ok]: EXPIRE with empty string as TTL should report an error (0 ms)
[ok]: SET with EX with big integer should report an error (0 ms)
[ok]: SET with EX with smallest integer should report an error (1 ms)
[ok]: GETEX with big integer should report an error (0 ms)
[ok]: GETEX with smallest integer should report an error (1 ms)
[ok]: EXPIRE with big integer overflows when converted to milliseconds (1 ms)
[ok]: PEXPIRE with big integer overflow when basetime is added (0 ms)
[ok]: EXPIRE with big negative integer (1 ms)
[ok]: PEXPIREAT with big integer works (0 ms)
[ok]: PEXPIREAT with big negative integer works (1 ms)
[ok]: All time-to-live(TTL) in commands are propagated as absolute timestamp in milliseconds in AOF (2064 ms)
[ok]: All TTL in commands are propagated as absolute timestamp in replication stream (37 ms)
[ok]: First server should have role slave after REPLICAOF (0 ms)
[ok]: For all replicated TTL-related commands, absolute expire times are identical on primary and replica (9 ms)

Logged crash report (pid 12908):
=== REDIS BUG REPORT START: Cut & paste starting from here ===
12908:S 08 Feb 2023 20:22:12.391 # === ASSERTION FAILED ===
12908:S 08 Feb 2023 20:22:12.391 # ==> blocked.c:624 'server.also_propagate.numops == 0' is not true

------ STACK TRACE ------

Backtrace:
src/redis-server 127.0.0.1:21613(handleClientsBlockedOnKeys+0x255)[0x562f67b01b65]
src/redis-server 127.0.0.1:21613(beforeSleep+0x125)[0x562f67a72555]
src/redis-server 127.0.0.1:21613(aeProcessEvents+0x8f)[0x562f67a6dfcf]
src/redis-server 127.0.0.1:21613(aeMain+0x1d)[0x562f67a6e53d]
src/redis-server 127.0.0.1:21613(main+0x332)[0x562f67a6a092]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7ff815e93c87]
src/redis-server 127.0.0.1:21613(_start+0x2a)[0x562f67a6a76a]

------ INFO OUTPUT ------
# Server
redis_version:7.0.8
redis_git_sha1:1c75ab06
redis_git_dirty:0
redis_build_id:5c85919483005628
redis_mode:standalone
os:Linux 5.15.79.1-microsoft-standard-WSL2 x86_64
arch_bits:64
monotonic_clock:POSIX clock_gettime
multiplexing_api:epoll
atomicvar_api:c11-builtin
gcc_version:7.5.0
process_id:12908
process_supervised:no
run_id:21e1a05d510f52af57938cfe5216ff4593c19b39
tcp_port:21613
server_time_usec:1675858932384193
uptime_in_seconds:0
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:14914548
executable:/home/jud/judeng/redis/src/redis-server
config_file:/home/jud/judeng/redis/./tests/tmp/redis.conf.12847.6
io_threads_active:0

# Clients
connected_clients:2
cluster_connections:0
maxclients:10000
client_recent_max_input_buffer:20480
client_recent_max_output_buffer:0
blocked_clients:0
tracking_clients:0
clients_in_timeout_table:0

# Memory
used_memory:1237992
used_memory_human:1.18M
used_memory_rss:12496896
used_memory_rss_human:11.92M
used_memory_peak:1275104
used_memory_peak_human:1.22M
used_memory_peak_perc:97.09%
used_memory_overhead:935924
used_memory_startup:858720
used_memory_dataset:302068
used_memory_dataset_perc:79.64%
allocator_allocated:1725576
allocator_active:2007040
allocator_resident:4866048
total_system_memory:8332824576
total_system_memory_human:7.76G
used_memory_lua:31744
used_memory_vm_eval:31744
used_memory_lua_human:31.00K
used_memory_scripts_eval:0
number_of_cached_scripts:0
number_of_functions:0
number_of_libraries:0
used_memory_vm_functions:32768
used_memory_vm_total:64512
used_memory_vm_total_human:63.00K
used_memory_functions:184
used_memory_scripts:184
used_memory_scripts_human:184B
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.16
allocator_frag_bytes:281464
allocator_rss_ratio:2.42
allocator_rss_bytes:2859008
rss_overhead_ratio:2.57
rss_overhead_bytes:7630848
mem_fragmentation_ratio:9.80
mem_fragmentation_bytes:11221880
mem_not_counted_for_evict:192
mem_replication_backlog:20508
mem_total_replication_buffers:20504
mem_clients_slaves:0
mem_clients_normal:56320
mem_cluster_links:0
mem_aof_buffer:192
mem_allocator:jemalloc-5.2.1
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Persistence
loading:0
async_loading:0
current_cow_peak:0
current_cow_size:0
current_cow_size_age:0
current_fork_perc:0.00
current_save_keys_processed:0
current_save_keys_total:0
rdb_changes_since_last_save:1
rdb_bgsave_in_progress:0
rdb_last_save_time:1675858932
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_saves:0
rdb_last_cow_size:0
rdb_last_load_keys_expired:0
rdb_last_load_keys_loaded:6
aof_enabled:1
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:0
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_rewrites:1
aof_rewrites_consecutive_failures:0
aof_last_write_status:ok
aof_last_cow_size:315392
module_fork_in_progress:0
module_fork_last_cow_size:0
aof_current_size:172
aof_base_size:172
aof_pending_rewrite:0
aof_buffer_length:0
aof_pending_bio_fsync:0
aof_delayed_fsync:0

# Stats
total_connections_received:2
total_commands_processed:53
instantaneous_ops_per_sec:32
total_net_input_bytes:2491
total_net_output_bytes:5618
total_net_repl_input_bytes:1674
total_net_repl_output_bytes:0
instantaneous_input_kbps:1.49
instantaneous_output_kbps:3.36
instantaneous_input_repl_kbps:1.00
instantaneous_output_repl_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:1
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:0
evicted_keys:0
evicted_clients:0
total_eviction_exceeded_time:0
current_eviction_exceeded_time:0
keyspace_hits:18
keyspace_misses:0
pubsub_channels:0
pubsub_patterns:0
pubsubshard_channels:0
latest_fork_usec:192
total_forks:1
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
total_active_defrag_time:0
current_active_defrag_time:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_error_replies:0
dump_payload_sanitizations:0
total_reads_processed:28
total_writes_processed:27
io_threaded_reads_processed:0
io_threaded_writes_processed:0
reply_buffer_shrinks:0
reply_buffer_expands:0

# Replication
role:slave
master_host:127.0.0.1
master_port:21611
master_link_status:up
master_last_io_seconds_ago:0
master_sync_in_progress:0
slave_read_repl_offset:2405
slave_repl_offset:2405
slave_priority:100
slave_read_only:0
replica_announced:1
connected_slaves:0
master_failover_state:no-failover
master_replid:d40a78f9d8d576edb0efe8233d1af3422b61edb8
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:2405
second_repl_offset:-1
repl_backlog_active:1
repl_backlog_size:1048576
repl_backlog_first_byte_offset:1101
repl_backlog_histlen:1305

# CPU
used_cpu_sys:0.005367
used_cpu_user:0.005131
used_cpu_sys_children:0.000000
used_cpu_user_children:0.000988
used_cpu_sys_main_thread:0.010093
used_cpu_user_main_thread:0.000000

# Modules

# Commandstats
cmdstat_get:calls=1,usec=3,usec_per_call=3.00,rejected_calls=0,failed_calls=0
cmdstat_set:calls=16,usec=58,usec_per_call=3.62,rejected_calls=0,failed_calls=0
cmdstat_replconf:calls=1,usec=1,usec_per_call=1.00,rejected_calls=0,failed_calls=0
cmdstat_restore:calls=2,usec=5,usec_per_call=2.50,rejected_calls=0,failed_calls=0
cmdstat_ping:calls=1,usec=1,usec_per_call=1.00,rejected_calls=0,failed_calls=0
cmdstat_pexpiretime:calls=17,usec=26,usec_per_call=1.53,rejected_calls=0,failed_calls=0
cmdstat_config|set:calls=2,usec=463,usec_per_call=231.50,rejected_calls=0,failed_calls=0
cmdstat_flushall:calls=1,usec=3252,usec_per_call=3252.00,rejected_calls=0,failed_calls=0
cmdstat_select:calls=2,usec=2,usec_per_call=1.00,rejected_calls=0,failed_calls=0
cmdstat_info:calls=1,usec=76,usec_per_call=76.00,rejected_calls=0,failed_calls=0
cmdstat_replicaof:calls=1,usec=168,usec_per_call=168.00,rejected_calls=0,failed_calls=0
cmdstat_pexpireat:calls=8,usec=9,usec_per_call=1.12,rejected_calls=0,failed_calls=0

# Errorstats

# Latencystats
latency_percentiles_usec_get:p50=3.007,p99=3.007,p99.9=3.007
latency_percentiles_usec_set:p50=2.007,p99=29.055,p99.9=29.055
latency_percentiles_usec_replconf:p50=1.003,p99=1.003,p99.9=1.003
latency_percentiles_usec_restore:p50=2.007,p99=3.007,p99.9=3.007
latency_percentiles_usec_ping:p50=1.003,p99=1.003,p99.9=1.003
latency_percentiles_usec_pexpiretime:p50=2.007,p99=2.007,p99.9=2.007
latency_percentiles_usec_config|set:p50=3.007,p99=460.799,p99.9=460.799
latency_percentiles_usec_flushall:p50=3260.415,p99=3260.415,p99.9=3260.415
latency_percentiles_usec_select:p50=1.003,p99=1.003,p99.9=1.003
latency_percentiles_usec_info:p50=76.287,p99=76.287,p99.9=76.287
latency_percentiles_usec_replicaof:p50=168.959,p99=168.959,p99.9=168.959
latency_percentiles_usec_pexpireat:p50=1.003,p99=2.007,p99.9=2.007

# Cluster
cluster_enabled:0

# Keyspace

------ CLIENT LIST OUTPUT ------
id=4 addr=127.0.0.1:37729 laddr=127.0.0.1:21613 fd=8 name= age=0 idle=0 flags=N db=9 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=20474 argv-mem=0 multi-mem=0 rbs=16384 rbp=0 obl=0 oll=0 omem=0 tot-mem=37632 events=r cmd=get user=default redir=-1 resp=2
id=6 addr=127.0.0.1:21611 laddr=127.0.0.1:54180 fd=9 name= age=0 idle=0 flags=M db=9 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=1287 argv-mem=0 multi-mem=0 rbs=16384 rbp=0 obl=0 oll=0 omem=0 tot-mem=18688 events=r cmd=flushall user=(superuser) redir=-1 resp=2

------ MODULES INFO OUTPUT ------

------ CONFIG DEBUG OUTPUT ------
client-query-buffer-limit 1gb
lazyfree-lazy-user-del no
lazyfree-lazy-expire no
sanitize-dump-payload no
io-threads-do-reads no
io-threads 1
lazyfree-lazy-eviction no
slave-read-only no
list-compress-depth 0
proto-max-bulk-len 512mb
lazyfree-lazy-server-del no
lazyfree-lazy-user-flush no
activedefrag no
repl-diskless-load disabled
repl-diskless-sync yes
replica-read-only no

------ FAST MEMORY TEST ------
12908:S 08 Feb 2023 20:22:12.392 # Bio thread for job type #0 terminated
12908:S 08 Feb 2023 20:22:12.392 # Bio thread for job type #1 terminated
12908:S 08 Feb 2023 20:22:12.392 # Bio thread for job type #2 terminated
*** Preparing to test memory region 562f67ec5000 (2301952 bytes)
*** Preparing to test memory region 562f69cbc000 (135168 bytes)
*** Preparing to test memory region 7ff81348a000 (8388608 bytes)
*** Preparing to test memory region 7ff813c8b000 (8388608 bytes)
*** Preparing to test memory region 7ff81448c000 (8388608 bytes)
*** Preparing to test memory region 7ff814c8d000 (8388608 bytes)
*** Preparing to test memory region 7ff815600000 (8388608 bytes)
*** Preparing to test memory region 7ff81625f000 (16384 bytes)
*** Preparing to test memory region 7ff81647e000 (16384 bytes)
*** Preparing to test memory region 7ff816e3e000 (32768 bytes)
*** Preparing to test memory region 7ff816e57000 (4096 bytes)
.O.O.O.O.O.O.O.O.O.O.O
Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

=== REDIS BUG REPORT END. Make sure to include from START to END. ===

       Please report the crash by opening an issue on github:

           http://github.com/redis/redis/issues

  If a Redis module was involved, please open in the module's repo instead.

  Suspect RAM error? Use redis-server --test-memory to verify it.

  Some other issues could be detected by redis-server --check-system

[exception]: Executing test client: I/O error reading reply.
I/O error reading reply
    while executing
"wait_for_condition 20 100 {
                [$replica get foo] eq {}
            } else {
                fail "can't delete expired the keys which is..."
    ("uplevel" body line 6)
    invoked from within
"uplevel 1 $code"
    (procedure "test" line 51)
    invoked from within
"test {expired keys which is created in writeable replicas should be delete} {
            $primary flushall
            $replica config set replica-re..."
    ("uplevel" body line 60)
    invoked from within
"uplevel 1 $code "
    (procedure "start_server" line 3)
    invoked from within
"start_server {tags {needs:repl external:skip}} {
        # Set the outer layer server as primary
        set primary [srv -1 client]
        set prima..."
    ("uplevel" body line 515)
    invoked from within
"uplevel 1 $code "
    (procedure "start_server" line 3)
    invoked from within
"start_server {tags {"expire"}} {
    test {EXPIRE - set timeouts multiple times} {
        r set x foobar
        set v1 [r expire x 5]
        set v2..."
    (file "tests/unit/expire.tcl" line 1)
    invoked from within
"source $path"
    (procedure "execute_test_file" line 4)
    invoked from within
"execute_test_file $data"
    (procedure "test_client_main" line 10)
    invoked from within
"test_client_main $::test_server_port "
Killing still running Redis server 12908
```

Reply:
I also commented out this line in the unstable branch, and geted the same crash

Reply:
@guybe7  I used `after 150` instead of `wait_for_condtion` as @enjoy-binbin suggested, and got the consistent result